### PR TITLE
[Modular probably?][DRAFT] Pain Component & Health System [DRAFT]

### PR DESCRIPTION
**Pull request by nevimer**
## About The Pull Request
**NOT DONE YET NOT EVEN CLOSE**
This PR serves as a draft for the pain and shock based modular health system I have been kinda sitting on the idea of for a while now. Currently, the code is very far from done, and all it does at the moment of opening this PR is give you a heart attack disease. 

## How This Contributes To The Skyrat Roleplay Experience

What if we moved away from raw damage numbers and made things like wounds matter more? 

## Proof of Testing

<details>
<summary>Screenshots/Videos</summary>
 https://cdn.discordapp.com/attachments/610319137012580383/1053184379351990383/dreamseeker_uocAVuSNBh.mp4
</details>

## Changelog

:cl:
add: Added pain and shock based health system. 
/:cl:

### CODE REVIEW

File: modular_skyrat/master_files/code/modules/health/health_mods.dm [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+	if(!istype(parent, /mob/living/carbon/human))
+		return COMPONENT_INCOMPATIBLE
+/datum/component/pain/RegisterWithParent()
+	pain_haver = parent
+	heart = pain_haver.getorgan(/obj/item/organ/internal/heart)
+	RegisterSignal(parent, COMSIG_LIVING_LIFE, .proc/pain_tick)
+/datum/component/pain/UnregisterFromParent()
+	pain_haver = null
+	heart = null
+	UnregisterSignal(parent, COMSIG_LIVING_LIFE)
```
Comment (softcerv) [inline modular_skyrat/master_files/code/modules/health/health_mods.dm:24]:
```suggestion
	UnregisterSignal(parent, COMSIG_LIVING_LIFE)
	
```

File: modular_skyrat/master_files/code/modules/health/health_mods.dm [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+	var/pain_score
+	var/obj/item/organ/internal/heart
+	var/list/last_wounds = list()
+/datum/component/pain/Initialize()
+	if(!istype(parent, /mob/living/carbon/human))
+		return COMPONENT_INCOMPATIBLE
+/datum/component/pain/RegisterWithParent()
+	pain_haver = parent
+	heart = pain_haver.getorgan(/obj/item/organ/internal/heart)
+	RegisterSignal(parent, COMSIG_LIVING_LIFE, .proc/pain_tick)
```
Comment (softcerv) [inline modular_skyrat/master_files/code/modules/health/health_mods.dm:20]:
```suggestion
	RegisterSignal(parent, COMSIG_LIVING_LIFE, .proc/pain_tick)
	
```

File: modular_skyrat/master_files/code/modules/health/health_mods.dm [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+#define MAGIC_PAIN_DIVISOR 10
+#define PAIN "pain"
+/datum/component/pain/
+	var/mob/living/carbon/human/pain_haver
+	var/pain_score
+	var/obj/item/organ/internal/heart
+	var/list/last_wounds = list()
+/datum/component/pain/Initialize()
+	if(!istype(parent, /mob/living/carbon/human))
+		return COMPONENT_INCOMPATIBLE
```
Comment (softcerv) [inline modular_skyrat/master_files/code/modules/health/health_mods.dm:16]:
```suggestion
		return COMPONENT_INCOMPATIBLE
		
```

File: modular_skyrat/master_files/code/modules/health/health_mods.dm [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+#define PAIN_WOUND_MINOR 30
+#define PAIN_WOUND_MODERATE 50
+#define PAIN_WOUND_MAJOR 80
+#define MAGIC_PAIN_DIVISOR 10
+#define PAIN "pain"
+/datum/component/pain/
+	var/mob/living/carbon/human/pain_haver
+	var/pain_score
+	var/obj/item/organ/internal/heart
+	var/list/last_wounds = list()
```
Comment (softcerv) [inline modular_skyrat/master_files/code/modules/health/health_mods.dm:13]:
```suggestion
	var/list/last_wounds = list()
	
```

File: modular_skyrat/master_files/code/modules/health/health_mods.dm [hunk 5]
```diff
@@ -0,0 +1,74 @@
+#define PAIN_CAP 500
+#define PAIN_MIN 0
+#define SHOCK_THRESH 50
+#define PAIN_WOUND_MINOR 30
+#define PAIN_WOUND_MODERATE 50
+#define PAIN_WOUND_MAJOR 80
+#define MAGIC_PAIN_DIVISOR 10
+#define PAIN "pain"
```
Comment (softcerv) [inline modular_skyrat/master_files/code/modules/health/health_mods.dm:8]:
```suggestion
#define PAIN "pain"

```

File: modular_skyrat/master_files/code/modules/health/health_mods.dm [hunk 6]
```diff
@@ -0,0 +1,74 @@
+#define PAIN_CAP 500
+#define PAIN_MIN 0
+#define SHOCK_THRESH 50
+#define PAIN_WOUND_MINOR 30
+#define PAIN_WOUND_MODERATE 50
+#define PAIN_WOUND_MAJOR 80
+#define MAGIC_PAIN_DIVISOR 10
```
Comment (softcerv) [inline modular_skyrat/master_files/code/modules/health/health_mods.dm:7]: could you write a comment or change the variable name to better explain how `MAGIC_PAIN_DIVISOR` works?

**Comment by YakumoChen** [comment]
I still have nightmares of BobMed...

Can you make like, some kind of design document for this?

File: modular_skyrat/master_files/code/modules/health/health_mods.dm [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+#define PAIN "pain"
+/datum/component/pain/
+	var/mob/living/carbon/human/pain_haver
```
Comment (MrMelbert) [inline modular_skyrat/master_files/code/modules/health/health_mods.dm:10]: don't bother, just typecast

**Review by MrMelbert — COMMENTED** [review]
suggestion (from someone who's made their own pain system): 

don't make this a component, especially not in your code base
people will want to read pain values and stuff, I 100% guarantee it. 

you can do this hella modularly with just a datum.

**Comment by nevimer** [comment]
> I still have nightmares of BobMed...
> 
> Can you make like, some kind of design document for this?

https://hackmd.io/c1owWVxhRvyYn8qli0FABw

Written a year ago, pretty out of date with my goals.

**Comment by YakumoChen** [comment]
> > I still have nightmares of BobMed...
> > Can you make like, some kind of design document for this?
> 
> https://hackmd.io/c1owWVxhRvyYn8qli0FABw
> 
> Written a year ago, pretty out of date with my goals.

can you update it then? A feature that replaces one of the core game mechanics should really have a design document attached so we know whats going on.

**Review by Zonespace27 — CHANGES_REQUESTED** [review]
To save you the pain (badum tssh) later, please update your design doc to your current goals in addition to being more detailed

**Comment by nevimer** [comment]
Sorry GitHub actions bot I'm too busy hunting undocumented diffs

**Comment by nevimer** [comment]
Zero chance of this happening, sorry folks.

**Pull request closed without merging**